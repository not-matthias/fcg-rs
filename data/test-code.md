---
cards-deck: code-test
---

## Testing whether the code works

```rust
fn main() {
    println!("Hello World");
}
```