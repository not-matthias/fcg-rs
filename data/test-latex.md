---
cards-deck: latex-test
---

## This is a question #card

$\top$
