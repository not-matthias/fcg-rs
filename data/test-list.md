---
cards-deck: list-test
---

## This is a question

- Item 1
- Item 2
- Item 3