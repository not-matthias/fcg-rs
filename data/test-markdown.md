## Header 2

Text 1

##### Header 5

Text 2

### Header 3

Text 3

---

# H1

A

## H2

B

## H2
## H2
Content
### H3
Content
### H3
Content
#### H4
Content
#### H4
Content
#### H4
Content

## H2
Content
### H3
Content
#### H4
Content