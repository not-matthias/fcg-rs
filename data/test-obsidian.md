---
cards-deck: obsidian-test
---

# Can this show images? 

Yes:

![[image.png]]
