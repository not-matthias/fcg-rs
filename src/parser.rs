use itertools::Itertools;
use petgraph::graph::NodeIndex;
use petgraph::prelude::EdgeRef;
use petgraph::{Direction, Graph};
use serde::Deserialize;

#[derive(Deserialize, Debug)]
pub struct FileHeader {
    #[serde(rename = "cards-deck")]
    pub cards_deck: String,
}

impl Default for FileHeader {
    fn default() -> Self {
        Self {
            cards_deck: "default".into(),
        }
    }
}

/// The struct representing a markdown header with content.
#[derive(Debug)]
pub struct HeaderWithContent {
    /// The header level.
    pub level: usize,

    /// The header of the block.
    pub header: String,

    /// The content of the header. This will be the back of the card.
    pub content: Vec<String>,
}

pub struct Parser {
    text: String,
}

impl Parser {
    pub fn new(text: String) -> Self {
        Self { text }
    }

    /// Tries to find a parent index for a heading.
    fn find_parent_index(
        graph: &Graph<HeaderWithContent, usize>,
        previous_index: NodeIndex,
        current_level: usize,
    ) -> Option<NodeIndex> {
        // We need to do the following:
        // - Check node at previous_index if it's already at the level above
        // - Go up all parents until we find the

        // Check if the parent is already correct
        //
        let is_parent_correct = |index| {
            if let Some(weight) = graph.node_weight(index) {
                if weight.level < current_level {
                    return Some(index);
                }
            }

            None
        };

        if let Some(index) = is_parent_correct(previous_index) {
            return Some(index);
        }

        // Find the correct parent by going up the chain
        //
        let mut current_index = previous_index;
        while let Some(edge) = graph
            .edges_directed(current_index, Direction::Incoming)
            .last()
        {
            current_index = edge.source();

            if let Some(index) = is_parent_correct(current_index) {
                return Some(index);
            }
        }

        None
    }

    fn parse_markdown(&mut self) -> Graph<HeaderWithContent, usize> {
        let mut graph = Graph::new();
        let mut previous_index: Option<NodeIndex> = None;
        let mut previous_level: Option<usize> = None;

        // Calculate where all the headers start
        //
        let header_indices = {
            let mut header_indices: Vec<_> = self
                .text
                .lines()
                .enumerate()
                .filter_map(|(index, line)| {
                    if line.contains('#') {
                        Some(index)
                    } else {
                        None
                    }
                })
                .collect();

            // We just say that the last line is also a heading so the last tuple has a end.
            //
            header_indices.push(self.text.lines().count());

            header_indices
        };

        // Iterate over the headers
        //
        for (index, next_index) in header_indices.into_iter().tuple_windows() {
            let header = self.text.lines().nth(index).unwrap_or_default();
            let level = header.chars().filter(|c| c.eq(&'#')).count();
            log::trace!("Header {:?} with level {:?}", header, level);

            let content = self
                .text
                .lines()
                .take(next_index)
                .skip(index)
                .filter(|s| !s.is_empty())
                .skip(1) // Skip header
                .collect::<Vec<_>>();
            if content.is_empty() {
                continue;
            }
            log::trace!("Content: {:?}", content);

            // Create the header
            //
            let header = HeaderWithContent {
                level,
                header: header.into(),
                content: content.into_iter().map(|s| s.to_string()).collect(),
            };

            if previous_index.is_some() && previous_level.is_some() {
                // Check if the new header is a sub-header
                // ```
                //       previous
                //     /         \
                //  HEADER     previous
                // ```
                if header.level > previous_level.unwrap() {
                    let node_index = graph.add_node(header);

                    if let Some(parent_index) =
                        Self::find_parent_index(&graph, previous_index.unwrap(), level)
                    {
                        graph.add_edge(parent_index, node_index, level);
                    }

                    previous_index = Some(node_index);
                    previous_level = Some(level);
                }
                // Check if the new header is an upper or equal header
                // ```
                //          HEADER
                //       /         \
                //  previous     previous
                // ```
                else if header.level <= previous_level.unwrap() {
                    let node_index = graph.add_node(header);

                    if let Some(parent_index) =
                        Self::find_parent_index(&graph, previous_index.unwrap(), level)
                    {
                        graph.add_edge(parent_index, node_index, level);
                    }

                    previous_index = Some(node_index);
                    previous_level = Some(level);
                }
            } else {
                previous_index = Some(graph.add_node(header));
                previous_level = Some(level);
            }
        }

        // std::fs::write("out.dot", format!("{:?}", Dot::with_config(&graph, &[])));

        graph
    }

    fn parse_yaml(&mut self) -> FileHeader {
        // The only acceptable input is this:
        // ```
        // ---
        // cards-deck: test
        // ```
        //
        // So we have to find the last `---`:
        //
        let minus_indices: Vec<_> = self
            .text
            .lines()
            .enumerate()
            .filter_map(|(index, line)| {
                if line.contains("---") {
                    Some(index)
                } else {
                    None
                }
            })
            .collect();

        if let Some(last_row) = minus_indices.last() {
            serde_yaml::from_str(&*self.text.lines().take(*last_row).join("\n")).unwrap_or_default()
        } else {
            Default::default()
        }
    }

    /// Parses and returns the header and markdown content as graph.
    pub fn parse(&mut self) -> (FileHeader, Graph<HeaderWithContent, usize>) {
        (self.parse_yaml(), self.parse_markdown())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parse_markdown() {
        let content = include_str!("../data/test-markdown.md").to_string();
        let mut parser = Parser::new(content);
        let content = parser.parse_markdown();

        insta::assert_debug_snapshot!(content);
    }

    #[test]
    fn parse_yaml() {
        let content = "---\ncards-deck: deck-name\n---".to_string();
        let mut parser = Parser::new(content);
        let content = parser.parse_yaml();

        insta::assert_debug_snapshot!(content);
    }
}
