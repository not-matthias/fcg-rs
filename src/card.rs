use crate::parser::HeaderWithContent;
use itertools::Itertools;
use petgraph::data::DataMap;
use petgraph::graph::NodeIndex;
use petgraph::prelude::EdgeRef;
use petgraph::{Direction, Graph};
use std::borrow::Cow;

#[derive(Debug)]
pub struct Card {
    pub(crate) front: String,
    pub(crate) back: String,
}

impl Card {
    fn header_context(graph: &Graph<HeaderWithContent, usize>, index: NodeIndex) -> Vec<String> {
        let mut context_data = Vec::new();
        let mut current_index = index;

        while let Some(edge) = graph
            .edges_directed(current_index, Direction::Incoming)
            .next()
        {
            current_index = edge.source();

            if let Some(context) = graph.node_weight(current_index).map(|h| h.header.clone()) {
                context_data.push(context.into());
            }
        }

        // We went from bottom up, so we have to reverse the vec to get the correct order.
        //
        context_data.reverse();

        context_data
    }

    pub fn new(graph: &Graph<HeaderWithContent, usize>, index: NodeIndex) -> Self {
        let node = graph.node_weight(index).unwrap();
        let context = Self::header_context(graph, index);

        // Remove the unneeded `#` from the context
        //
        let context = context
            .into_iter()
            .map(|c| c.replace("# ", "").replace("#", ""))
            .join(" > ");

        let back = node.content.join("\n").trim().to_string();
        let back = comrak::markdown_to_html(&back, &comrak::ComrakOptions::default());

        Self {
            front: format!("{} > {}", context, node.header.replace('#', "").trim()).into(),
            back: back,
        }
    }
}
