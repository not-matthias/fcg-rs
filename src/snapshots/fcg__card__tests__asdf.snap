---
source: src/card.rs
expression: converted
---

        ### Wie kann man von \(\mathbb R ^{\mathbb N}\) ausgehend einen Vektorraum konstruieren?
